examples/shader/texture_binding_array.rs

```diff
@@ -0,0 +1,151 @@
+//! A shader that binds several textures onto one
+//! `binding_array<texture<f32>>` shader binding slot and sample non-uniformly.
+
+use bevy::{
+    prelude::*,
+    reflect::TypeUuid,
+    render::{
+        render_asset::RenderAssets,
+        render_resource::{AsBindGroupError, PreparedBindGroup, *},
+        renderer::RenderDevice,
+        texture::FallbackImage,
+    },
+};
+use std::num::NonZeroU32;
+
+fn main() {
+    App::new()
+        .add_plugins(DefaultPlugins.set(ImagePlugin::default_nearest()))
+        .add_plugin(MaterialPlugin::<BindlessMaterial>::default())
+        .add_startup_system(setup)
+        .run();
+}
+
+const MAX_TEXTURE_COUNT: usize = 16;
+const TILE_ID: [usize; 16] = [
+    19, 23, 4, 33, 12, 69, 30, 48, 10, 65, 40, 47, 57, 41, 44, 46,
+];
+
+fn setup(
+    mut commands: Commands,
+    mut meshes: ResMut<Assets<Mesh>>,
+    mut materials: ResMut<Assets<BindlessMaterial>>,
+    asset_server: Res<AssetServer>,
+) {
+    commands.spawn(Camera3dBundle {
+        transform: Transform::from_xyz(2.0, 2.0, 2.0).looking_at(Vec3::new(0.0, 0.0, 0.0), Vec3::Y),
+        ..Default::default()
+    });
+
+    // load 16 textures
+    let textures: Vec<_> = TILE_ID
+        .iter()
+        .map(|id| {
+            let path = format!("textures/rpg/tiles/generic-rpg-tile{:0>2}.png", id);
+            asset_server.load(path)
+        })
+        .collect();
+
+    // a cube with multiple textures
+    commands.spawn(MaterialMeshBundle {
+        mesh: meshes.add(Mesh::from(shape::Cube { size: 1.0 })),
+        material: materials.add(BindlessMaterial { textures }),
+        ..Default::default()
+    });
+}
+
+#[derive(Debug, Clone, TypeUuid)]
+#[uuid = "8dd2b424-45a2-4a53-ac29-7ce356b2d5fe"]
+struct BindlessMaterial {
+    textures: Vec<Handle<Image>>,
+}
+
+impl AsBindGroup for BindlessMaterial {
```

Comment:
would be great to get this into the derive macro. i don't think we need to block on that but lets add an issue at least when this is merged.

Reply:
Yes. It's worth noting that `AsBindGroup` has a limitation that `bind_group_layout()` is a static function, and that's why there needs a constant `MAX_TEXTURE_COUNT` here.